src/helpers.c

```diff
@@ -191,88 +191,80 @@ int system_argv(char *const argv[])
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
+	pid_t childpid = fork();
+	if (childpid == -1) {
+		LOG(NULL, "Failed to create child process", "");
+		return -1;
+	} else if (childpid == 0) {
+		int pipefd[2];
+		if (pipe(pipefd) == -1) {
+			LOG(NULL, "Failed to create a pipe", "");
+			assert(0);
 		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
+
+		pid_t lhspid = fork();
+		if (lhspid == -1) {
+			LOG(NULL, "Failed to fork LHS process", "");
+			return -1;
+		} else if (lhspid == 0) {
+			if (dup2(pipefd[1], STDOUT_FILENO) == -1) {
 				LOG(NULL, "Could not redirect stdout", "");
 				assert(0);
 			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
+			execvp(*lhscmd, lhscmd);
+			LOG(NULL, "LHS command failed", "");
+			assert(0);
+		} else {
+			pid_t rhspid = fork();
+			if (rhspid == -1) {
+				LOG(NULL, "Failed to fork RHS process", "");
+				return -1;
+			} else if (rhspid == 0) {
+				if (dup2(pipefd[0], STDIN_FILENO) == -1) {
+					LOG(NULL, "Could not redirect stdin", "");
+					assert(0);
+				}
+				execvp(*rhscmd, rhscmd);
+				LOG(NULL, "LHS command failed", "");
 				assert(0);
+			} else {
+				int lhsresult = wait_process_terminate(lhspid);
+				int rhsresult = wait_process_terminate(rhspid);
+				close(pipefd[0]);
+				close(pipefd[1]);
+				if (lhsresult == 0 && rhsresult == 0) {
+					exit(EXIT_SUCCESS);
+				} else {
+					if (lhsresult != 0) {
+						exit(lhsresult);
+					} else {
+						exit(rhsresult);
+					}
+				}
 			}
-			close(newstderr);
 		}
-
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
-		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
-		assert(0);
 	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
-
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
-			assert(0);
-		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
-			assert(0);
-		}
-
-		if (status != 0) {
-			char *cmdline = NULL;
-
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
-		}
+		int childresult = wait_process_terminate(childpid);
+		return childresult;
 	}
-
-	return status;
 }
 
-int system_argv_pipe(char *const argvp1[], int stdinp1, int stderrp1,
-		     char *const argvp2[], int stdoutp2, int stderrp2)
+int wait_process_terminate(pid_t pid)
 {
-	int statusp2;
-	int pipefd[2];
-
-	if (pipe(pipefd)) {
-		LOG(NULL, "Failed to create a pipe", "");
-		return -1;
+	int status;
+	do {
+		if (waitpid(pid, &status, 0) == -1) {
+			LOG(NULL, "Failed to wait for PID", "");
+			return -1;
+		}
+	} while (!WIFEXITED(status) && !WIFSIGNALED(status));
+	if (WIFEXITED(status)) {
```

Comment:
Not a great fan of this return code stuff. It makes it hard to tell if tar returned say 6 or tar was killed with a SIGABRT.

Either just return the status value as is, or do what ksh does and return 256+WTERMSIG(status) in the signal case. (Note bash retuns 128+WTERMSIG(status) which doesn't allow you to distinguish half the valid return codes, but is better than nothing).